internal/model/keys.go

```diff
@@ -2,85 +2,157 @@
 package model
 
 import (
+	"crypto/rand"
 	"database/sql"
 	"database/sql/driver"
 	"fmt"
+	"math/big"
 	"reflect"
-	"regexp"
+	"strconv"
 	"strings"
-
-	"github.com/google/uuid"
-	"github.com/oklog/ulid/v2"
 )
 
-func NewRunnerKey() RunnerKey                      { return RunnerKey(ulid.Make()) }
-func ParseRunnerKey(key string) (RunnerKey, error) { return parseKey[RunnerKey](key) }
+func NewRunnerKey(hostname string, port string) RunnerKey {
+	suffix, err := rand.Int(rand.Reader, big.NewInt(10000))
+	if err != nil {
+		panic(err)
+	}
+	return keyType[runnerKey]{
+		Hostname: hostname,
+		Port:     port,
+		Suffix:   int(suffix.Int64()),
+	}
+}
+func NewLocalRunnerKey(suffix int) RunnerKey {
+	return keyType[runnerKey]{
+		Suffix: suffix,
+	}
+}
+func ParseRunnerKey(key string) (RunnerKey, error) { return parseKey[RunnerKey](key, true) }
 
 type runnerKey struct{}
 type RunnerKey = keyType[runnerKey]
 
-func NewControllerKey() ControllerKey                      { return ControllerKey(ulid.Make()) }
-func ParseControllerKey(key string) (ControllerKey, error) { return parseKey[ControllerKey](key) }
+func NewControllerKey(hostname string, port string) ControllerKey {
+	suffix, err := rand.Int(rand.Reader, big.NewInt(10000))
+	if err != nil {
+		panic(err)
+	}
+	return keyType[controllerKey]{
+		Hostname: hostname,
+		Port:     port,
+		Suffix:   int(suffix.Int64()),
+	}
+}
+func NewLocalControllerKey(suffix int) ControllerKey {
+	return keyType[controllerKey]{
+		Suffix: suffix,
+	}
+}
+func ParseControllerKey(key string) (ControllerKey, error) { return parseKey[ControllerKey](key, true) }
 
 type controllerKey struct{}
 type ControllerKey = keyType[controllerKey]
 
-var uuidRe = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)
+func parseKey[KT keyType[U], U any](key string, includesKind bool) (KT, error) {
+	// Expected style: [<kind>-]<host>-<port>-<suffix> or [<kind>-]<suffix>
+
+	components := strings.Split(key, "-")
+	if includesKind {
+		//
+		if len(components) == 0 {
+			return KT{}, fmt.Errorf("expected a prefix for key: %s", key)
+		}
+		kind := kindFromType[U]()
+		if components[0] != kind {
+			return KT{}, fmt.Errorf("unexpected prefix for key: %s", key)
+		}
+		components = components[1:]
+	}
 
-func parseKey[KT keyType[U], U any](key string) (KT, error) {
-	var zero KT
-	kind := kindFromType[U]()
 	switch {
-	case strings.HasPrefix(key, kind):
-		ulid, err := ulid.Parse(key[len(kind):])
+	case len(components) == 1:
+		//style: [<kind>-]<suffix>
+
+		suffix, err := strconv.Atoi(components[len(components)-1])
```

Comment:
When you switch to hex encoding you can use `strconv.ParseInt()` with a base of 16 here.